sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala

```diff
@@ -65,3 +93,25 @@ case class StringAsDataTypeException(
     fieldName: String,
     fieldValue: String,
     dataType: DataType) extends RuntimeException()
+
+/**
+ * No-stacktrace equivalent of `QueryExecutionErrors.cannotParseJSONFieldError`.
+ * Used for code control flow in the parser without overhead of creating a full exception.
+ */
+case class CannotParseJSONFieldException(
+    fieldName: String,
+    fieldValue: String,
+    jsonType: JsonToken,
+    dataType: DataType) extends RuntimeException() {
+  override def getStackTrace(): Array[StackTraceElement] = new Array[StackTraceElement](0)
+  override def fillInStackTrace(): Throwable = this
+}
+
+/**
+ * No-stacktrace equivalent of `QueryExecutionErrors.emptyJsonFieldValueError`.
+ * Used for code control flow in the parser without overhead of creating a full exception.
```

Comment:
So we still use exceptions in the control flow, but use a trick to avoid the stacktrace overhead?

Can we avoid using stacktrace in the control flow? or at lease not in the critical code path. e.g. `fieldConverters` should return null directly if the input is empty string and `enablePartialResults` is true.

Reply:
That is correct, we still use exceptions to report any issues in parsing.

Throwing exceptions allows us to propagate errors across several functions. Without this, we would have to include checks in the code for every result. Also, we would have to capture the exception itself to be reported to the user. If we decided to go with returning some class to indicate failed parsing, then we would essentially be implementing exceptions. Exceptions are not that much slower, most of the time is spent on building the stacktrace which we only need when the very first exception is thrown.